### PR TITLE
Implement PowerQuery support for SentinelOne

### PR DESCRIPTION
**Pull request by jholtmann**
This PR implements PowerQuery support for the SentinelOne product. 

**Benefits of PowerQuery (PQ) over Deep Visibility (DV):**
- Lack of a 1req/min rate limit allows product to perform a separate query per search criteria, resulting in more useful output
- Ability to perform a query per search criteria helps reduce the impact of the 20k search result limit

**Discussion Points:**
- Which mode should be the default, PQ or DV? PQ provides faster and more accurate search results, but I am not sure if it is available across all S1 consoles. This PR currently sets PQ as the default and allows falling back to DV with the `--dv` product flag.
- PQ does not respect Site ID filters unless the parent Account ID is also included in the query. This means that if the user provides both an Account ID and a nested Site ID, we cannot assume that only the Account ID should be queried (as we currently do in DV mode). This PR automatically adds the Account ID when only a Site ID is provided by the user. This logic difference between DV and PQ may be confusing to users.
- Having both PQ and DV implementations makes the code somewhat harder to understand due to frequent branching logic. If we determine that PowerQuery can be assumed to be globally available, we may want to consider removing DV support in the future.

**Testing:**
- I have tested PQ and DV searches with various definition files and haven't noticed any issues.

### CODE REVIEW

File: products/sentinel_one.py [hunk 1]
```diff
@@ -414,63 +464,60 @@ def nested_process_search(self, tag: Tag, criteria: dict, base_query: dict):
...
 
-        try:
-            # queries with certain operators can be combined into a more compact query format
+        if self._pq:
+            query_text = list[Tuple[Tag, str]]()
 
+            for tag, queries in self._queries.items():
+                for query in queries:
+                    full_query = f'{query.parameter} {query.operator} {query.search_value}'
+                    query_text.append((tag, full_query))
```
Comment (rc-csmith) [inline products/sentinel_one.py:503]: When I only specify the `--query` param and don't use definition files, the query isn't built correctly

```suggestion
                    if query.full_query is not None:
                        query_text.append((tag, query.full_query))
                    else:
                        full_query = f'{query.parameter} {query.operator} {query.search_value}'
                        query_text.append((tag, full_query))
```

Error log:
```
[2023-02-23 09:14:38,187] [DEBUG   ] [surveyor.s1                         ] [common.py           :54  ] Authenticated
[2023-02-23 09:14:38,189] [DEBUG   ] [surveyor                            ] [help.py             :36  ] Running Custom Query: src.process.name contains "svchost.exe"
[2023-02-23 09:14:38,189] [DEBUG   ] [surveyor.s1                         ] [help.py             :36  ] Built Query: src.process.name contains "svchost.exe"
[2023-02-23 09:14:38,189] [DEBUG   ] [surveyor.s1                         ] [sentinel_one.py     :713 ] Entered get_results
[2023-02-23 09:14:38,189] [DEBUG   ] [surveyor.s1                         ] [sentinel_one.py     :717 ] Executing additional _process_queries
[2023-02-23 09:14:38,190] [DEBUG   ] [surveyor.s1                         ] [sentinel_one.py     :564 ] Query params: {'accountIds': ['REDACTED'], 'fromDate': 1677163478189, 'toDate': 1677165278189, 'limit': 20000, 'query': 'None None None | group count() by endpoint.name, src.process.user, src.process.image.path, src.process.cmdline, src.process.name, src.process.publisher, url.address, tgt.file.internalName, event.time, site.id, site.name'}
```

File: products/sentinel_one.py [hunk 2]
```diff
@@ -479,89 +526,179 @@ def _process_queries(self):
...
+            query_response.raise_for_status()
 
-                query_id = body['data']['queryId']
-                self.log.info(f'Query ID is {query_id}')
+            query_id = body['data']['queryId']
+            self.log.info(f'Query ID is {query_id}')
 
-                events = self._get_dv_events(query_id)
-                self.log.debug(f'Got {len(events)} events')
+            events = self._get_dv_events(query_id, p_bar=p_bar, cancel_event=cancel_event)
```
Comment (rc-csmith) [inline products/sentinel_one.py:585]: When testing, I came across this issue when the script checked the status of a PowerQuery, 503 errors got thrown regardless of if the query is valid or if there's results being returned. Adding this section appeared to clear up the errors. I'm wondering if the API doesn't allow you to ping a PowerQuery's status if the search is finished in the initial call?

```suggestion
            if self._pq and body['data']['status'] == 'FINISHED': # If using PQ, the results can be returned immediately
                events = body['data']['data']
            else:
                events = self._get_dv_events(query_id, p_bar=p_bar, cancel_event=cancel_event)
```

**Review by rc-csmith — CHANGES_REQUESTED** [review]
Just a few changes since I came across a couple bugs. Otherwise, works like a charm and is so much better than DV!!

**Comment by xC0uNt3r7hr34t** [comment]
PowerQuery is still in Beta. I would recommend that DV be set to default until PowerQuery is released to GA and more stable. I agree with all the other benefits. If we want to keep PowerQuery as the default for the future to limit code changes, then we need to make it very obvious in documentation and help guide that PowerQuery is default.

**Comment by rc-csmith** [comment]
Because of the benefits of using PQ (more accurate results and faster runtime), I think we should go ahead and keep it as defaul. But, as @xC0uNt3r7hr34t said, we'll need to make sure documentation and release notes are in place when the changes go live so users are aware.

**Pull request merged**